### PR TITLE
SALTO-2802: fetch two section with the same name in zendesk guid

### PR DESCRIPTION
**Pull request by edenhassid**
_fetch two section with the same name in zendesk guid_

---

_Additional context for reviewer_
None

---
_Release Notes_: 
None

---
_User Notifications_: 
None

### CODE REVIEW

File: packages/zendesk-adapter/src/adapter.ts [hunk 1]
```diff
@@ -206,11 +210,22 @@ const zendeskGuideEntriesFunc = (
       brandPaginatorResponseValues.forEach(response => {
         const responseEntryName = typesConfig[typeName].transformation?.dataField
         if (responseEntryName === undefined) {
-          return undefined
+          return
         }
-        return (response[responseEntryName] as Value[]).forEach(instanceType => {
+        (response[responseEntryName] as Value[]).forEach(instanceType => {
```
Comment (netama) [inline packages/zendesk-adapter/src/adapter.ts:215]: nit (for the existing code): these are not `instanceType`s, please rename them to `entry` or something like that (and also use that in the new code)

File: packages/zendesk-adapter/src/adapter.ts [hunk 2]
```diff
@@ -206,11 +210,22 @@ const zendeskGuideEntriesFunc = (
...
           instanceType.brand = brandInstance.value.id
         })
+        if (responseEntryName === SECTIONS_TYPE_NAME) {
+          (response[responseEntryName] as Value[]).forEach(instanceType => {
+            if (instanceType.parent_section_id !== null) {
```
Comment (netama) [inline packages/zendesk-adapter/src/adapter.ts:220]: weird, is it really `null` here and not `undefined`?

File: packages/zendesk-adapter/src/adapter.ts [hunk 3]
```diff
@@ -206,11 +210,22 @@ const zendeskGuideEntriesFunc = (
...
+              instanceType.direct_parent = instanceType.parent_section_id
+              instanceType.parent_type = 'section'
```
Comment (netama) [inline packages/zendesk-adapter/src/adapter.ts:222]: if we are relying on these later, please use shared consts

File: packages/zendesk-adapter/src/adapter.ts [hunk 4]
```diff
@@ -206,11 +210,22 @@ const zendeskGuideEntriesFunc = (
...
         const responseEntryName = typesConfig[typeName].transformation?.dataField
         if (responseEntryName === undefined) {
-          return undefined
+          return
         }
-        return (response[responseEntryName] as Value[]).forEach(instanceType => {
+        (response[responseEntryName] as Value[]).forEach(instanceType => {
           instanceType.brand = brandInstance.value.id
         })
+        if (responseEntryName === SECTIONS_TYPE_NAME) {
```
Comment (netama) [inline packages/zendesk-adapter/src/adapter.ts:218]: let's add a comment here explaining why these are needed

File: packages/zendesk-adapter/src/filters/help_center_parent_to_section.ts [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+  if (elem.value.parent_section_id !== undefined) {
+    elem.value.direct_parent = elem.value.parent_section_id
+    elem.value.parent_type = 'section'
+  } else {
+    elem.value.direct_parent = elem.value.category_id
+    elem.value.parent_type = 'category'
+  }
+}
+
+const filterCreator: FilterCreator = ({ client, config }) => ({
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/help_center_parent_to_section.ts:44]: please add a docstring

File: packages/zendesk-adapter/src/filters/help_center_parent_to_section.ts [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+import { addRemovalChangesId } from './help_center_section_and_category'
+
+export const SECTION_TYPE_NAME = 'section'
+
+const deleteParentFields = (elem: InstanceElement): void => {
+  delete elem.value.direct_parent
+  delete elem.value.parent_type
+}
+
+const addParentFields = (elem: InstanceElement): void => {
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/help_center_parent_to_section.ts:34]: can we reuse this function in `zendeskGuideEntriesFunc`?

File: packages/zendesk-adapter/src/filters/help_center_parent_to_section.ts [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+      .filter(change => getChangeData(change).elemID.typeName === SECTION_TYPE_NAME)
+      .map(getChangeData)
+      .forEach(deleteParentFields)
+  },
+  deploy: async (changes: Change<InstanceElement>[]) => {
+    const [parentChanges, leftoverChanges] = _.partition(
+      changes,
+      change => SECTION_TYPE_NAME === getChangeData(change).elemID.typeName
+    )
+    addRemovalChangesId(parentChanges)
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/help_center_parent_to_section.ts:56]: looking at this now, I see I missed that this is changing something in the global context. I think we will need to separate this list by brand, no?

File: packages/zendesk-adapter/src/filters/help_center_parent_to_section.ts [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    const deployResult = await deployChanges(
+      parentChanges,
+      async change => {
+        await deployChange(change, client, config.apiDefinitions, ['translations', 'parent_section_id'])
+      }
+    )
+    // need to deploy separately parent_section_id if exists
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/help_center_parent_to_section.ts:63]: it's great that this is documented, might be good to elaborate on why

File: packages/zendesk-adapter/src/group_change.ts [hunk 9]
```diff
@@ -55,8 +56,15 @@ const recurseIntoInstanceChangeToGroupId: ChangeIdFunction = async change => {
 const typeNameChangeGroupId: ChangeIdFunction = async change =>
   getChangeData(change).elemID.typeName
 
+const SectionChangeGroupId: ChangeIdFunction = async change =>
```
Comment (netama) [inline packages/zendesk-adapter/src/group_change.ts:59]: :+1: please add an explanation why these cannot be grouped together (the dependency on the parent's id)

**Review by netama — COMMENTED** [review]
looks great! added a few comments (most are nits)

File: packages/zendesk-adapter/src/adapter.ts (hunk 2 above)
Comment (edenhassid) [reply packages/zendesk-adapter/src/adapter.ts:220]: yes it was weird to me too

File: packages/zendesk-adapter/src/group_change.ts [hunk 10]
```diff
@@ -55,8 +57,16 @@ const recurseIntoInstanceChangeToGroupId: ChangeIdFunction = async change => {
 const typeNameChangeGroupId: ChangeIdFunction = async change =>
   getChangeData(change).elemID.typeName
 
+// sections need to be grouped separately as there are dependencies with 'parent_section_id'
+const SectionChangeGroupId: ChangeIdFunction = async change =>
```
Comment (netama) [inline packages/zendesk-adapter/src/group_change.ts:61]: nit: start lowecase for "variable" names
```suggestion
const sectionChangeGroupId: ChangeIdFunction = async change =>
```

**Review by netama — APPROVED** [review]
added a couple of questions, pre-approving (didn't do a full pass but I think we discussed most of these :) )

File: packages/zendesk-adapter/src/filters/help_center_parent_to_section.ts [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+          const ParentSectionInstance = new InstanceElement(
+            getChangeData(change).elemID.name,
+            await getChangeData(change).getType(),
+            {
+              id: getChangeData(change).value.id,
+              parent_section_id: getChangeData(change).value.parent_section_id,
+            }
+          )
+          await deployChange(
+            toChange({ before: ParentSectionInstance, after: ParentSectionInstance }),
```
[the remark this answers is not in the document]
Comment (edenhassid) [reply packages/zendesk-adapter/src/filters/help_center_parent_to_section.ts:81]: yes, because I need `toChange` to turn it to a modification change

**Pull request merged**